Enh: `file` fields: `metadata`

martin-rueegg opened the pull request:
Adds `metadata` field to `File` table/model.

### What kind of change does this PR introduce?

- Feature

### Does this PR introduce a breaking change?

- No

### The PR fulfills these requirements

- [x] It's submitted to the `develop` branch, _not_ the `master` branch if no hotfix
- [x] All [tests](#issuecomment-new) are passing
- [ ] New/updated tests are included
- [x] Changelog was modified

If adding a **new feature**, the PR's description includes:
- [ ] A convincing reason for adding this feature (to avoid wasting your time, it's best to open a suggestion issue first and wait for approval before working on it)

**Other information:**

Depends on/includes:
- #6590

luke- commented on protected/humhub/libs/StdClass.php:
```diff
@@ -0,0 +1,412 @@
+use Traversable;
+use yii\helpers\ArrayHelper;
+
+class StdClass extends \stdClass implements ArrayAccess, Stringable, SeekableIterator, Countable, Serializable
```
This class seems very powerful. What are the advantages to the regular `\stdClass`. Do we really need  these additional features as well? Are they used?

luke- commented on protected/humhub/libs/StdClassConfig.php:
```diff
@@ -0,0 +1,26 @@
+
+use ReflectionObject;
+
+class StdClassConfig
```
Is this used? How does it work?

martin-rueegg replied:
In order to not occupy any property names, some additional data can be stored in an attached "config" object. This is accessible through the `static::config()` function.

It is now only available in the `StdClassConfigurable` extension of `StdClass`

It stores the default value (that will be returned if a non-existing property was read. It also allows to lock the property set (so no new properties can be added).

luke- commented on protected/humhub/modules/file/models/File.php:
```diff
@@ -8,56 +8,74 @@
  */
 class File extends FileCompat
 {
+    public const WELL_KNOWN_METADATA_UPLOAD_HASH = 'file._upload.hash';
```
We should hold these `const` back for now.
Maybe it is useful to create an own `FileMetadata` class later, which then contains such constants.

martin-rueegg replied:
Its funny that I was thinking the same, but I had to dash ...

`Metadata` _is_ actually in the `File` model-specific. As such, I've moved the over there, but only in the follow-up PR that actually fills the metadata during upload:
- #6554

luke- commented on protected/humhub/modules/file/models/File.php:
```diff
@@ -112,12 +139,37 @@ public function behaviors()
+     * @noinspection PhpMissingReturnTypeInspection
+     * @noinspection ReturnTypeCanBeDeclaredInspection
+     */
+    public function transactions()
```
We are close to the last beta release and stable release. I don't want to enable transactions in this stage.

martin-rueegg replied:
disabled in https://github.com/humhub/humhub/pull/6590/commits/3480969ae6a0410ae5aa5ac78f3eed6a0d64b98b. See https://github.com/humhub/humhub/pull/6590#discussion_r1335063469 for more info.

luke- commented on protected/humhub/modules/file/models/File.php:
```diff
@@ -112,12 +139,37 @@ public function behaviors()
+    {
+        $value = parent::__get($name);
+
+        if ($name === 'metadata' && !$value instanceof Metadata) {
```
What about using this here?
```php
if ($name === 'metadata') {
   return $this->getMetadata();
}
```

martin-rueegg replied:
Yes, that works, too!

Your solution is certainly easier to understand. My solution was maybe minimally more efficient on successive calls ...

luke- commented on protected/humhub/modules/file/models/File.php:
```diff
@@ -8,56 +8,74 @@
- * @property \humhub\modules\user\models\User $createdBy
- * @property \humhub\modules\file\components\StorageManager $store
+ * @property StorageManager $store
  * @property FileHistory[] $historyFiles
```
We should also add the `metadata` property here.

martin-rueegg replied:
Absolutely. Got lost in all the cherry-picking, rebasing, and merging! 🤣

luke- commented on protected/humhub/modules/file/models/File.php:
```diff
@@ -240,24 +313,60 @@ public function canDelete($userId = null)
+     *
+     * @return File
+     */
+    public function setMetadata($metadata): File
```
TBD

martin-rueegg replied:
Alternatively, this could be removed entirely and make the property read-only.

- To set a value, one would need to do:
   ``` php
   $model->getMetadata()->property = 'value';
   $model->metadata->property = 'value';
   ```
   You find more examples here:
   https://github.com/humhub/humhub/blob/78df7d89e17ac95106fd7afa41ed326f6dd61df8/protected/humhub/modules/file/models/File.php#L51-L62
- With the current implementation of `setMetadata()`, the above could be like so:
   ``` php
   $model->setMetadata('property' => 'value');
   
   // this would delete all the data within the metadata object
   $model->setMetadata(null);
   $model->metadata = null;
   ```

luke- reviewed (changes requested):
Some first thoughts

martin-rueegg commented:
I've added some documentation as why this is better than `\stdClass` (see below).

For me the main advantages are
- security during unserialization
- structured content

My suggestion is to require the top-level property of the metadata to be a module ID. The given model 
- needs to be enabled
- has to register the namespace
- can then define the structure of it's content, by providing a class that is used to instantiate it's property

E.g., as I further suggest, the `File` module would use the top-level properties to correspond with the given variant of a file. Such as
- `_org` [^1]
- `_upload`
- `_preview`
- etc ...
Each of these properties would result in an object that stores file-information, e.g.,
- hash
- mime-type (for variants)
- size
- maybe dimensions

[^1]: I'd like to change that to `_original` at a later point. I find `_org` very confusing as it is not a common abbreviation for "original", but for "organization".

This would result in a structure like

- `$metadata`
  - `file`
    - `_upload`
      - `hash`
      - `size`
      - `mime-type`
      - `original_filename`
    - `_draft` 
      - `hash`
      - `size`
      - `mime-type`
  - `onlyoffice`
    - `key` 


You can find a sample implementation [here](https://github.com/humhub/humhub/compare/develop...metaworx:humhub:enh/file-preliminary-fields-metadata-1?expand=1), a branch that builds on this PR.

--

Here I've added some documentation:

- https://github.com/humhub/humhub/blob/9a15bcc60833c6e98947a00c35c39e9d38d500f9/protected/humhub/libs/StdClass.php#L36-L59
- https://github.com/humhub/humhub/blob/9a15bcc60833c6e98947a00c35c39e9d38d500f9/protected/humhub/libs/StdClassConfigurable.php#L20-L27

luke- reviewed (approved):
@martin-rueegg Thanks for the PR. I merge it as discussed in the Beta.2 However, I'm not sure if the `StdClass*` is my favorite solution here, but lets discuss this later.
